src/JsonSchema/Validator.php

```diff
@@ -60,6 +60,8 @@ public function validate(&$value, $schema = null, $checkMode = null)
         // add provided schema to SchemaStorage with internal URI to allow internal $ref resolution
         if (is_object($schema) && property_exists($schema, 'id')) {
             $schemaURI = $schema->id;
+        } elseif (is_array($schema) && array_key_exists('id', $schema)) {
+            $schemaURI = $schema['id'];
         } else {
             $schemaURI = SchemaStorage::INTERNAL_PROVIDED_SCHEMA_URI;
         }
```

Comment:
I believe I've learned something more about the codebase today. These changes should be done respecting the `CHECK_MODE_TYPE_CAST`. Doing either a loose type check or a strict type check. I believe the following would do so (but requires some testing as well)

```suggestion
        $schemaURI = SchemaStorage::INTERNAL_PROVIDED_SCHEMA_URI;
        if ($this->factory->getTypeCheck()::propertyExists($schema, 'id')) {
            $schemaURI = $this->factory->getTypeCheck()::propertyGet($schema, 'id');
        }
```